theano/tensor/opt.py

```diff
@@ -5841,3 +5841,26 @@ def local_add_mul_fusion(node):
 def local_grad_clip(node):
     if isinstance(node.op, theano.gradient.GradClip):
         return node.inputs
+
+@register_canonicalize
+@register_stabilize
+@register_specialize
+@gof.local_optimizer([T.Alloc])
+def local_reduce_double_alloc(node):
+    # Alloc(Alloc(m, x, 1), x, y) -> Alloc(m, x, y)
+    if not isinstance(node.op, T.Alloc):
+        return False
+    if not node.inputs[0].owner or not isinstance(node.inputs[0].owner.op, T.Alloc):
+        return False
+    inputs_outer = node.inputs
+    inputs_inner = node.inputs[0].owner.inputs
+    if len(inputs_outer) != len(inputs_inner):
+        return False
```

Comment:
code updated, and also a rebase to clear up old commits.

On Thu, Jul 2, 2015 at 4:46 PM, Frédéric Bastien <notifications@github.com>
wrote:

> In theano/tensor/opt.py
> <https://github.com/Theano/Theano/pull/3082#discussion_r33821044>:
>
> > @@ -5841,3 +5841,26 @@ def local_add_mul_fusion(node):
> >  def local_grad_clip(node):
> >      if isinstance(node.op, theano.gradient.GradClip):
> >          return node.inputs
> > +
> > +@register_canonicalize
> > +@register_stabilize
> > +@register_specialize
> > +@gof.local_optimizer([T.Alloc])
> > +def local_reduce_double_alloc(node):
> > +    # Alloc(Alloc(m, x, 1), x, y) -> Alloc(m, x, y)
> > +    if not isinstance(node.op, T.Alloc):
> > +        return False
> > +    if not node.inputs[0].owner or not isinstance(node.inputs[0].owner.op, T.Alloc):
> > +        return False
> > +    inputs_outer = node.inputs
> > +    inputs_inner = node.inputs[0].owner.inputs
> > +    if len(inputs_outer) != len(inputs_inner):
> > +        return False
>
> T.alloc(T.alloc(m, y, 1, 1), x, y, z, w) should get optimzed by this opt.
> The inner alloc broadcast on y, the second on x, z and w.
>
> The if above will cause this opt do not cover this case. You also need to
> modify code bellow the if to make it work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3082/files#r33821044>.
>